### PR TITLE
[Product Creation AI] Update prompt drafting to reduce unexpected AI response

### LINKED ISSUE #11188: Product Creation with AI: Details Not Visible in Preview Screen on Xcode 15 and iOS 17
I'm experiencing an issue while using product creation with AI on Xcode 15 and iOS 17. Specifically, the Name, Description and Short Description of the product are not visible in the preview screen. 

However, when I compile the app with Xcode 14 (and iOS 16), everything works fine and all the product details are visible as expected.

### Steps to Reproduce:

1. Run the app compiled with Xcode 15.
2. Navigate to the product creation feature and use AI to create a new product.
3. Notice that Name, Description and Short Description in the preview are hidden.

### Additional Information:

This issue seems to be specific to Xcode 15 and iOS 17, as the problem does not occur when using Xcode 14 and iOS 16.

### PR DESCRIPTION
**Pull request by selanthiraiyan**
Closes: #11188 

## Description

Updates the product creation AI prompt to reduce the possibility of receiving unexpected extra response from AI. 

Changes
- Remove newline characters and empty spaces added for human readability and use a dictionary and JSON encoding to create the prompt. 
- Add a sentence to the prompt asking to send only JSON and nothing else.

In my testing, I see AI responding with unexpected responses sometimes. But we handle that by showing the error alert and asking the user to Retry. i.e. We don't treat that as a success case. Please report if you see any other issues. 

## Testing instructions
- Create a JN store with Jetpack AI plugin
- Login into the app
- Navigate to the Products tab
- Tap "+" -> "Create a product with AI"
- Follow the steps to enter the product name and features
- Validate that the AI generates the product details as expected
- Try the AI creation flow a few more times to generate diverse products and validate that you don't see an empty screen like in https://github.com/woocommerce/woocommerce-ios/issues/11188#issuecomment-1839949721 

## Screenshots

https://github.com/woocommerce/woocommerce-ios/assets/524475/86165972-40ec-4814-926e-01754b1c3479

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Networking/Networking/Remote/GenerativeContentRemote.swift [hunk 1]
```diff
@@ -199,74 +199,62 @@ private extension GenerativeContentRemote {
                            categories: [ProductCategory],
                            tags: [ProductTag],
                            token: JWToken) async throws -> AIProduct {
+        let input = [
+            "You are a WooCommerce SEO and marketing expert, perform in-depth research about the product " +
+            "using the provided name, keywords and tone.",
+            "Your response should be in JSON format and don't send anything extra.",
```
Comment (itsmeichigo) [inline Networking/Networking/Remote/GenerativeContentRemote.swift:205]: IMO we should be more specific about this to avoid the cases where [some fields are nested](https://github.com/woocommerce/woocommerce-ios/issues/11188#issuecomment-1841537724):

```suggestion
            "Your response should be in a valid JSON format. Each field should contain a value in String, no nested dictionaries are accepted.",
```

File: Networking/Networking/Remote/GenerativeContentRemote.swift [hunk 2]
```diff
@@ -199,74 +199,62 @@ private extension GenerativeContentRemote {
...
             }()
 
-            return "shipping: {" +
-                            weightJson +
-                            dimensionsJson +
-                    "}, "
+            return ["name": "The name of the product, in the ISO language code ```\(language)```",
+                    "description": "Product description of around 100 words long in a ```\(tone)``` tone, "
+                    + "in the ISO language code ```\(language)```",
+                    "short_description": "Product's short description, in the ISO language code ```\(language)```",
```
Comment (itsmeichigo) [inline Networking/Networking/Remote/GenerativeContentRemote.swift]: These prompts seem to not be clear enough that sometimes the contents are nested in the language code. How about rewording this:

```suggestion
            return ["name": "The name of the product, written in the language with ISO code ```\(language)```",
                    "description": "Product description of around 100 words long in a ```\(tone)``` tone, "
                    + "written in the language with ISO code ```\(language)```",
                    "short_description": "Product's short description, written in the language with ISO code ```\(language)```",
```

**Review by itsmeichigo — COMMENTED** [review]
While this works for me when testing, I think we should add extra validation for the response. 

When decoding name, description, and short description, we are [allowing](https://github.com/woocommerce/woocommerce-ios/blob/283cbfd48d2f862c910f987e004c1fde8bd0522f/Networking/Networking/Model/Product/AIProduct.swift#L86-L88) these fields to be empty if the decoding fails, which results in these fields being [empty](https://github.com/woocommerce/woocommerce-ios/issues/11188#issuecomment-1839949721).

I'd suggest we add some enhancements:
- Require that at least the description is available. If decoding fails, we treat the result as a failure and let the user try again.
- If the title cannot be decoded, we can use the title given by the user.
- If the short description cannot be decoded, we can hide that field from the screen instead of leaving it empty.

The above are just my suggestions, we can discuss further or share the idea with the team to align with everyone.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/183">17.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Networking/Networking/Remote/GenerativeContentRemote.swift (hunk 2 above)
Comment (selanthiraiyan) [reply Networking/Networking/Remote/GenerativeContentRemote.swift]: I agree. Done in https://github.com/woocommerce/woocommerce-ios/pull/11853/commits/24bbeb33713d3ae1779ced829e523bc6a4b54e43

File: Networking/Networking/Remote/GenerativeContentRemote.swift (hunk 1 above)
Comment (selanthiraiyan) [reply Networking/Networking/Remote/GenerativeContentRemote.swift:205]: Unfortunately, we cannot use this as we expect a nested JSON object for `shipping` field. 

Did you face the nested issue again after this PR's changes?

**Comment by selanthiraiyan** [comment]
Those are all valid suggestions, @itsmeichigo!

I have made the following changes.
- Throw an error if the description cannot be parsed from the AI response. This way, we will show the error alert from which the user can retry. 
- Use the entered product name if no product name is returned from AI. 
- Update the wording in the prompt to mention the ISO language code clearly. (As per PR comment)
- Hide the short description field if empty.

FYI, I will inform the Android team about the changes after we finalize and merge the PR. 

The PR is ready for another look. 🙇

File: Networking/Networking/Remote/GenerativeContentRemote.swift (hunk 1 above)
Comment (itsmeichigo) [reply Networking/Networking/Remote/GenerativeContentRemote.swift:205]: I understand, please ignore my suggestion for asking no nested dictionaries then. I cannot consistently reproduce the nested issue, I hope the update for language prompts helps with that.

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Add Product/AddProductWithAI/ProductDetailPreviewViewModelTests.swift [hunk 3]
```diff
@@ -601,6 +601,143 @@ final class ProductDetailPreviewViewModelTests: XCTestCase {
...
+        await viewModel.generateProductDetails()
+
+        // Then
+        let generatedProduct = try XCTUnwrap(viewModel.generatedProduct)
+        XCTAssertEqual(generatedProduct.name, "Pen")
+    }
+
+    // MARK: Short description view
+
+    func test_short_description_view_is_shown_if_shortDescription_is_not_empty() async throws {
```
Comment (itsmeichigo) [inline WooCommerce/WooCommerceTests/ViewRelated/Products/Add Product/AddProductWithAI/ProductDetailPreviewViewModelTests.swift]: Nit: `throws` is redundant here (similar to other tests below).

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Add Product/AddProductWithAI/ProductDetailPreviewViewModelTests.swift [hunk 4]
```diff
@@ -601,6 +601,143 @@ final class ProductDetailPreviewViewModelTests: XCTestCase {
...
+
+        mockProductActions(aiGeneratedProductResult: .success(product))
+        mockProductTagActions()
+        mockProductCategoryActions()
+
+        // When
+        await viewModel.generateProductDetails()
+
+        // Then
+        XCTAssertFalse(viewModel.shouldShowShortDescriptionView)
```
Comment (itsmeichigo) [inline WooCommerce/WooCommerceTests/ViewRelated/Products/Add Product/AddProductWithAI/ProductDetailPreviewViewModelTests.swift]: This should be true right?

**Review by itsmeichigo — APPROVED** [review]
Thanks for the updates, this looks good to me now 👍

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Add Product/AddProductWithAI/ProductDetailPreviewViewModelTests.swift (hunk 4 above)
Comment (selanthiraiyan) [reply WooCommerce/WooCommerceTests/ViewRelated/Products/Add Product/AddProductWithAI/ProductDetailPreviewViewModelTests.swift]: Right. Done in https://github.com/woocommerce/woocommerce-ios/pull/11853/commits/cbc66fcdb9fc79eeb6475042ce046cb79d179881.

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Add Product/AddProductWithAI/ProductDetailPreviewViewModelTests.swift (hunk 3 above)
Comment (selanthiraiyan) [reply WooCommerce/WooCommerceTests/ViewRelated/Products/Add Product/AddProductWithAI/ProductDetailPreviewViewModelTests.swift]: Removed in https://github.com/woocommerce/woocommerce-ios/pull/11853/commits/69fdf0a1bf633f23e3a8228851dec8b50708b063
